### PR TITLE
Rework of cloudnet's module loading

### PR DESCRIPTION
**Pull request by derklaro**
This pull request includes:

- [ ] breaking changes
- [X] no breaking changes

**This pull requests is on hold as we are waiting for our module gradle plugin to be deployed to the gradle plugin portal in order to add it to all existing modules before merging.**

### Changes made to the repository:

The complete module system was reworked and is now working smoother and is completely documented.

### Documentation of test results:

Modules are still working and starting as expected without any changes made to them.

### CODE REVIEW

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/DefaultModuleTaskEntry.java [hunk 1]
```diff
@@ -16,36 +16,96 @@
...
+public class DefaultModuleTaskEntry implements IModuleTaskEntry {
 
-  private final IModuleWrapper moduleWrapper;
+  /**
+   * A simple format which allows unique method identifiers. Format: {@code <declaring class>@<method name>()}
+   */
+  protected static final String METHOD_SIGNATURE_FORMAT = "%s@%s()";
 
-  private final ModuleTask taskInfo;
+  protected final Method jlrMethod;
```
Comment (0utplay) [inline cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/DefaultModuleTaskEntry.java:31]: What does jlr mean

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleProvider.java [hunk 2]
```diff
@@ -20,53 +20,229 @@
...
+   * @see ModuleLifeCycle
+   */
+  @NotNull
+  @Unmodifiable Collection<IModuleWrapper> getModules();
+
+  /**
+   * Get all loaded, started, stopped modules provided by this provider which have the specific given group.
+   *
+   * @param group the group id of the modules to get.
+   * @return all loaded, started, stopped modules provided by this provider which have the specific given group.
```
Comment (0utplay) [inline cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleProvider.java]: Should we perhaps inform that this collection is not modifiable, apart from the annotation?

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java [hunk 3]
```diff
@@ -18,47 +18,160 @@
...
-
-  IModuleWrapper loadModule();
-
-  IModuleWrapper startModule();
-
-  IModuleWrapper stopModule();
+  /**
+   * Get all module tasks which were detected in the main class of the module.
+   *
+   * @return all module tasks which were detected in the main class of the module.
```
Comment (0utplay) [inline cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java]: Should we perhaps inform that this map is not modifiable, apart from the annotation?

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java [hunk 4]
```diff
@@ -18,47 +18,160 @@
...
+  @Unmodifiable Map<ModuleLifeCycle, List<IModuleTaskEntry>> getModuleTasks();
 
-  IModuleWrapper unloadModule();
+  /**
+   * Get all modules this module is depending on.
+   *
+   * @return all modules this module is depending on.
+   */
+  @NotNull
+  @Unmodifiable Set<ModuleDependency> getDependingModules();
```
Comment (0utplay) [inline cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java:52]: see above

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java [hunk 5]
```diff
@@ -18,47 +18,160 @@
...
+   *
+   * @return the module configuration on which base the module was created.
+   */
+  @NotNull ModuleConfiguration getModuleConfiguration();
+
+  /**
+   * @deprecated Use {@link #getModuleConfiguration()} instead - same result but unwrapped.
+   */
+  @Deprecated
+  default JsonDocument getModuleConfigurationSource() {
```
Comment (0utplay) [inline cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java:87]: Add @ ScheduledForRemoval as some people ignore @ Deprecated

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java [hunk 6]
```diff
@@ -18,47 +18,160 @@
...
+   * @return the same instance of this class, for chaining.
+   * @see #getModuleTasks()
+   * @see ModuleLifeCycle#canChangeTo(ModuleLifeCycle)
+   * @see IModuleProvider#notifyPreModuleLifecycleChange(IModuleWrapper, ModuleLifeCycle)
+   */
+  @NotNull IModuleWrapper unloadModule();
+
+  /**
+   * @deprecated Use {@link #getDataDirectory()} instead.
+   */
```
Comment (0utplay) [inline cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java:144]: maybe remove it now?

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java [hunk 7]
```diff
@@ -18,47 +18,160 @@
...
+   * Get the uri from where the module was loaded.
+   *
+   * @return the uri from where the module was loaded.
+   */
+  @NotNull URI getUri();
 
+  /**
+   * @deprecated Use {@link ModuleConfiguration#getRepos()} instead.
+   */
+  @Deprecated
```
Comment (0utplay) [inline cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java:174]: @ ScheduledForRemoval

**Review by 0utplay — CHANGES_REQUESTED** [review]

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/DefaultModuleTaskEntry.java (hunk 1 above)
Comment (derklaro) [reply cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/DefaultModuleTaskEntry.java:31]: java.lang.reflect

Comment (0utplay) [reply cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/DefaultModuleTaskEntry.java:31]: I think something like reflectMethod is more appropriate

File: cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java (hunk 6 above)
Comment (derklaro) [reply cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java:144]: I'm currently not a fan of removing these as the deprecated annotation was added in 3.4... I think we go with 3.6 to have "buffer" versions.

Comment (0utplay) [reply cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java:144]: If thats the case you should add @ ScheduledForRemoval?

Comment (derklaro) [reply cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java:144]: The provblem is that no method which return java.io.File was annotated with @ ScheduledForRemval so I'm not willing to add it here...

**Comment by 0utplay** [comment]
LGTM

**Review by juliarn — APPROVED** [review]

**Review by 0utplay — APPROVED** [review]

**Pull request merged**